snapcraft/internal/pluginhandler/stage_package_grammar/_processor.py

```diff
@@ -0,0 +1,138 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+
+from ._errors import StagePackageSyntaxError
+
+
+def process_grammar(grammar, project_options, ubuntu):
+    """Process stage packages grammar and extract packages to actually stage.
+
+    Arguments:
+        grammar: List containing stage-packages grammar.
+        project_options: Instance of ProjectOptions to use to determine stage
+                         packages.
+        ubuntu: repo.Ubuntu instance used for checking package validity.
+
+    Returns:
+        Set of packages to stage.
+    """
+
+    from ._on import OnStatement
+    from ._try import TryStatement
+
+    packages = set()
+    on_clause_pattern = re.compile(r'\Aon\s+')
+    try_clause_pattern = re.compile(r'\Atry\Z')
+    else_clause_pattern = re.compile(r'\Aelse\Z')
+    else_fail_pattern = re.compile(r'\Aelse\s+fail\Z')
+    statements = _StatementCollection()
+    statement = None
+
+    for section in grammar:
+        try:
+            if else_fail_pattern.match(section):
+                _handle_else(statement, None)
+            else:
+                # If it's just a string, add it to the package set.
+                packages.add(section)
+        except TypeError:
+            # Alright, it wasn't an unconditional stage package, so it must be
+            # a dict.
```

Comment:
Honestly I'd prefer to use `isinstance`, but I thought duck typing was more pythonic. Happy to change.